packages/pwa-kit-react-sdk/src/ssr/universal/hooks/index.js

```diff
@@ -70,3 +71,45 @@ export const useOrigin = ({fromXForwardedHeader = false}) => {
     }
     return APP_ORIGIN
 }
+
+/**
+ * Blocks the navigation to run a provided function whenever there is a new page being navigated to
+ * The function must return false to unblock the navigation
+ *
+ * @param {function} func
+ * @returns {boolean} State that is set to true during blocking, false otherwise.
+ */
+export const useBlockNavigation = (func) => {
+    const {block, push, location} = useHistory()
+    const lastLocation = useRef()
+    const [isBlocked, setIsBlocked] = useState(false)
+    const funcRef = useRef()
+    const abortControllerRef = useRef(new AbortController())
+    funcRef.current = func
+    useEffect(() => {
+        const unblock = block((location, action) => {
+            if (location?.pathname !== lastLocation.current?.pathname && funcRef.current) {
+                lastLocation.current = location
+                abortControllerRef.current.abort()
+                abortControllerRef.current = new AbortController()
+                ;(async () => {
+                    setIsBlocked(true)
+                    if (
+                        !(await funcRef.current(
+                            location,
+                            action,
+                            abortControllerRef.current.signal
+                        ))
+                    ) {
+                        setIsBlocked(false)
+                        unblock()
+                        push(location)
+                    }
+                })()
+                return false
+            }
+        })
```

Comment:
`useEffect` can return a cleanup function if needed. In our case, there is something to clean up. 

```suggestion
        })
        return () => unblock()
```